Log out other sessions including offline sessions on password change

ahus1 opened the pull request:
Closes #38850

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

ahus1 committed:
Log out other sessions including offline sessions on password change

Closes #38850

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
@keycloak/core-clients - this is the first of possibly multiple PRs to handle offline sessions better. 
I'd be happy if one of you can review this one. 

While I consider this an important bug that might eventually be backported, I'm happy to hear other people's opinions on weather this should be backported to KC 26.0 for example.

mposolda reviewed (commented):
@ahus1 The question whether the "Logout all" should trigger only "online" sessions or also "offline" sessions is an outstanding issue, which we discussed many times on various occasions.

It is probably better to logout also offline sessions when user has checkbox "Sign out all sessions" checked. However it could be considered breaking change as it it not compatible with the current behaviour.

Also IMO not everyone wants this as a desired behaviour as offline sessions are primarily used for long-standing tasks. For example client application using offline tokens to perform some periodic tasks every night, which client application is doing even if user is already inactive and logged-out for a 2 months. And not sure why changing the password of the user would expect to logout all those periodic tasks, which the same user may started few weeks ago...

The backchannel logout specification also has something, which indicates that logout may not automatically logout offline session https://openid.net/specs/openid-connect-backchannel-1_0.html#BCActions . Even though the use-case from the specification is a bit different, but it can be seen that this is tricky topic...

Maybe the best option is to show something to the users/administrators whether the "Logout all" should revoke only online sessions, or also offline sessions. However not sure how to make it usable without not confuse user too much? I've found old thread when I suggested to do something for the admin console, but this is also not yet done AFAIK (even if some things work differently than mentioned in that email thread): https://lists.jboss.org/pipermail/keycloak-dev/2018-June/010907.html

ahus1 commented:
@mposolda - thank you for your inputs here, and the concerns for breaking behavior. The link you provided is IMHO about the admin console, while this is about the user when they for example update their password. See below for a screenshot hat is presented to an end-user. That was missing in the original issue, and I now added it both in the issue and now here for better context.

The link about the backchannel logout is IMHO also different from this scenario: If I am in a browser, and log out in my browser, I want to log out all sessions that are connected to this browser session (or regular user session in our case). Offline sessions are not connected to the browser session, so they IMHO shouldn't be affected. This is not about logging out from a single browser, but it is about the checkbox "Sign out from other devices" explicitly. 

The text is "Sign out from other devices", which would include then IMHO also mobile apps which usually have offline tokens. 

The situation where you change your password and want to sign-out from other sessions is IMHO if you suspect that someone knows your password, and you now want to make sure there are no more open sessions. In such a case, it wouldn't make much sense to log out some sessions, while keeping others. 

I suspect it goes beyond this screen to explain users the difference between regular and offline sessions. So adding more text is not a solution IMHO.

We could argue to change the default of the check-box to be un-checked, still having it checked is "secure by default", and most people wouldn't understand the checkbox if they are changing their password. So I would keep it as is.

So I think this PR now fixes the behavior how it should have been given the UI presented to the user and the assumed intent of the user. 

I agree it might be breaking to some, still it is IMHO ok to break a behavior if it fixes a bug. We could bring in an SPI option to revert to the old behavior at least until we publish 27.0 if we think backwards compatibility is an issue here. We would target 26.3 here, and we could put it to the migration guide, but I would prefer not to have an SPI to revert to the old behavior. But I would be ok to add it if you or others think this is the way to go. 

![image](https://github.com/user-attachments/assets/6081620b-b95f-4d4b-95d5-c9303300e715)

mposolda commented:
@ahus1 Yes, I was aware of the screen which you mentioned in the issue and the use-case of `Sign out from other devices` .

I was pointing the admin console and backchannel specification just for the illustration that "Logout all" is a bit tricky thing in general and it is questionable if it should logout also offline sessions or not...

But I do agree with your point that it is rather better to logout also offline sessions for this use-case. To me, it would makes sense to document this in the migration guide and also introduce backwards compatibility option, which we document that the option is planned to be removed in 27.0. Maybe something similar we recently did for the "JWT client authentication" in keycloak 26.2 : https://www.keycloak.org/docs/latest/upgrading/index.html#jwt-client-authentication-aligned-with-the-latest-oidc-specification . If lots of people have concerns with changing the default behaviour, we can revisit and make option supported even after 27.0 though. WDYT?

ahus1 committed:
Rework by introducing a feature to fall back to the old behavior

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Updating changes

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
@mposolda - I've now added a feature flag so admins can revert to the original behavior. 

I didn't use an SPI option as contrary to the example you provided above this is not about different clients that might not support this behavior, but about the behavior presented to the users. Adding more options to configure Keycloak adds to the complexity, so this deprecated option should be removed to reduce the cost of added complexity. 

I've also updated the migration guide accordingly. 

Please re-review when you have the time. Thanks!

mposolda reviewed (approved):
@ahus1 LGTM, Thanks
